experimental/pattern-matching.md

```diff
@@ -0,0 +1,119 @@
+# [Pattern Matching](proposal-pattern-matching)
+
+See also [AST examples](ast-examples) for this proposal.
+
+## MatchExpression
+
+```js
+interface MatchExpression <: Expression {
+    type: "MatchExpression";
+    discriminant: Expression;
+    id: Pattern | null;
+    clause: [ MatchClause ]
+}
+```
+
+## WhenClause
+```js
+interface MatchClause <: Node {
+    type: "MatchClause";
+    test: MatchPattern | null;
+    guard: Expression | null;
+    id: Pattern | null;
+    consequent: BlockStatement
+}
+```
+
+If `test` is `null`, `id` must be `null`.
+
+## MatchPattern
+```js
+interface MatchPattern <: Node {}
+```
+
+## ArrayMatchPattern
+```js
+interface ArrayMatchPattern <: MatchPattern {
+    type: "ArrayMatchPattern";
+    elements: [ MatchPattern ];
+}
+```
+
+## ObjectMatchPattern
+```js
+interface AssignmentMatchProperty <: Property {
+    type: "Property"; // inherited
+    value: MatchPattern;
+    kind: "init";
+    method: false;
+}
+
+interface ObjectMatchPattern <: MatchPattern {
+    type: "ObjectMatchPattern";
+    properties: [ AssignmentMatchProperty ];
+}
+```
+
+## RestMatchElement
```

Comment:
The `ArrayMatchPattern`, `ObjectMatchPattern` and `RestMatchElement` are essentially `ArrayPattern`, `ObjectPattern` and `RestElement` with `Pattern` replaced by `MatchPattern`.